File: AndroidApp/app/src/main/java/com/jamespfluger/alexadevicefinder/activities/NameActivity.java
package com.jamespfluger.alexadevicefinder.activities;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.jamespfluger.alexadevicefinder.R;
import com.jamespfluger.alexadevicefinder.models.CommonData;
import com.jamespfluger.alexadevicefinder.utilities.PreferencesManager;

public class NameActivity extends AppCompatActivity {
    private PreferencesManager preferencesManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preferencesManager = new PreferencesManager(getApplicationContext());

        setContentView(R.layout.activity_name);
        initializeUi();
    }

    private void initializeUi() {
        final EditText deviceNameField = findViewById(R.id.deviceNameField);
        final Button deviceNameContinueButton = findViewById(R.id.deviceNameContinueButton);
        // TODO: remove this before going live
        deviceNameField.setText("James");

        deviceNameField.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) {
                    InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(NameActivity.INPUT_METHOD_SERVICE);

                    if (inputMethodManager != null) {
                        inputMethodManager.hideSoftInputFromWindow(v.getWindowToken(), 0);
                    }
                } else {
                    TextView deviceNameErrorDescription = findViewById(R.id.deviceNameErrorDescription);
                    deviceNameErrorDescription.setVisibility(View.INVISIBLE);
                }
            }
        });

        deviceNameContinueButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                validateDeviceNameField(deviceNameField);
            }
        });
    }

    private void validateDeviceNameField(final EditText deviceNameField) {
        String deviceName = deviceNameField.getText().toString();

        if (deviceNameField.getText().length() == 0) {
            TextView deviceNameErrorDescription = findViewById(R.id.deviceNameErrorDescription);
            deviceNameErrorDescription.setText("You must provide a device name before continuing.");
            deviceNameErrorDescription.setVisibility(View.VISIBLE);

            final Animation errorAnimation = AnimationUtils.loadAnimation(this, R.anim.shake);
            deviceNameField.startAnimation(errorAnimation);
        } else if (!CommonData.names.contains(deviceName.toLowerCase())) {
            DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    switch (which) {
                        case DialogInterface.BUTTON_POSITIVE:
                            preferencesManager.setDeviceName(deviceNameField.getText().toString());
                            switchToActivity(OtpActivity.class);
                            break;
                        case DialogInterface.BUTTON_NEGATIVE:
                            break;
                    }
                }
            };
            AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this);
            alertBuilder.setTitle("Warning");
            alertBuilder.setMessage("The name \"" + deviceName + "\" is not a common first name.\nWhile not required, the skill works best if you use a first name.\n\nAre you sure you want to continue?");
            alertBuilder.setPositiveButton("Yes", dialogClickListener);
            alertBuilder.setNegativeButton("No", dialogClickListener);
            alertBuilder.show();

        } else {
            preferencesManager.setDeviceName(deviceNameField.getText().toString());
            switchToActivity(OtpActivity.class);
        }
    }

    private void switchToActivity(Class<?> newActivity) {
        Intent newIntent = new Intent(this, newActivity);
        startActivity(newIntent);
        finish();
    }
}

File: MobileApp/AlexaDeviceFinder/AlexaDeviceFinder.Android/Activities/NameActivity.cs
using System;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Views;
using Android.Views.Animations;
using Android.Views.InputMethods;
using Android.Widget;
using DeviceFinder.Droid.Models;
using DeviceFinder.Droid.Utilities;
using static Android.Views.View;

namespace DeviceFinder.Droid.Activities
{
    public class NameActivity : Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {
        private PreferencesManager preferencesManager;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            preferencesManager = new PreferencesManager(this.ApplicationContext);

            base.SetContentView(Resource.Layout.activity_name);
            InitializeUi();
        }

        private void InitializeUi()
        {
            EditText deviceNameField = (EditText)this.FindViewById(Resource.Id.deviceNameField);
            Button deviceNameContinueButton = (Button)this.FindViewById(Resource.Id.deviceNameContinueButton);
            // TODO: remove this before going live
            deviceNameField.SetText("James", TextView.BufferType.Normal);

            deviceNameField.FocusChange += DeviceNameFieldFocusChange;
            deviceNameContinueButton.Click += ContinueButtonClick;
        }

        private void ContinueButtonClick(object sender, EventArgs args)
        {
            EditText deviceNameField = this.FindViewById(Resource.Id.deviceNameField) as EditText;
            string deviceName = deviceNameField.Text;

            if (deviceNameField.Text.Length == 0)
            {
                TextView deviceNameErrorDescription = this.FindViewById(Resource.Id.deviceNameErrorDescription) as TextView;
                deviceNameErrorDescription.Text = "You must provide a device name before continuing.";
                deviceNameErrorDescription.Visibility = ViewStates.Visible;

                Animation errorAnimation = AnimationUtils.LoadAnimation(this, Resource.Animation.shake);
                deviceNameField.StartAnimation(errorAnimation);
            }
            else if (!CommonData.Names.Contains(deviceName.ToLowerInvariant()))
            {
                AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this);
                alertBuilder.SetTitle("Warning");
                alertBuilder.SetMessage("The name \"" + deviceName + "\" is not a common first name.\nWhile not required, the skill works best if you use a first name.\n\nAre you sure you want to continue?");
                alertBuilder.SetPositiveButton("Yes", YesAction);
                alertBuilder.SetNegativeButton("No", YesAction);
                alertBuilder.Show();
            }
            else
            {
                preferencesManager.SetDeviceName(deviceNameField.Text);
                SwitchToActivity(typeof(OtpActivity));
            }
        }

        private void YesAction(object sender, DialogClickEventArgs args)
        {
            EditText deviceNameField = this.FindViewById(Resource.Id.deviceNameField) as EditText;
            preferencesManager.SetDeviceName(deviceNameField.Text);
            SwitchToActivity(typeof(OtpActivity));
        }

        private void SwitchToActivity(Type newActivity)
        {
            Intent newIntent = new Intent(this, newActivity);
            this.StartActivity(newIntent);
            this.Finish();
        }

        private void DeviceNameFieldFocusChange(object sender, FocusChangeEventArgs args)
        {
            if (!args.HasFocus)
            {
                InputMethodManager inputMethodManager = (InputMethodManager)this.GetSystemService(Context.InputMethodService);

                if (inputMethodManager != null)
                {
                    inputMethodManager.HideSoftInputFromWindow(((View)sender).ApplicationWindowToken, 0);
                }
            }
            else
            {
                TextView deviceNameErrorDescription = this.FindViewById(Resource.Id.deviceNameErrorDescription) as TextView;
                deviceNameErrorDescription.Visibility = ViewStates.Invisible;
            }
        }
    }
}